File: src/test/java/com/onezero/oop/settlement/SettlementTest.java
package com.onezero.oop.settlement;

import java.util.UUID;

public class SettlementTest {
    public static void main(String[] args) {
        allowsNormalTransition();
        blocksInvalidTransition();
        allowsCorrectionRequestAfterConfirmed();
        blocksConfirmingFailedSettlement();
        exposesCurrentStatus();
        System.out.println("SettlementTest passed");
    }

    private static void allowsNormalTransition() {
        Settlement settlement = new Settlement(UUID.randomUUID());

        settlement.transitionTo(SettlementStatus.AGGREGATING);
        settlement.transitionTo(SettlementStatus.VALIDATING);
        settlement.transitionTo(SettlementStatus.CONFIRMED);

        assertEquals(SettlementStatus.CONFIRMED, settlement.status());
    }

    private static void blocksInvalidTransition() {
        Settlement settlement = new Settlement(UUID.randomUUID());

        assertThrows(InvalidTransitionException.class, () -> settlement.transitionTo(SettlementStatus.CONFIRMED));
    }

    private static void allowsCorrectionRequestAfterConfirmed() {
        Settlement settlement = new Settlement(UUID.randomUUID());

        settlement.transitionTo(SettlementStatus.AGGREGATING);
        settlement.transitionTo(SettlementStatus.VALIDATING);
        settlement.transitionTo(SettlementStatus.CONFIRMED);
        settlement.transitionTo(SettlementStatus.CORRECTION_REQUESTED);

        assertEquals(SettlementStatus.CORRECTION_REQUESTED, settlement.status());
    }

    private static void blocksConfirmingFailedSettlement() {
        Settlement settlement = new Settlement(UUID.randomUUID());

        settlement.transitionTo(SettlementStatus.FAILED);

        assertThrows(InvalidTransitionException.class, () -> settlement.transitionTo(SettlementStatus.CONFIRMED));
    }

    private static void exposesCurrentStatus() {
        Settlement settlement = new Settlement(UUID.randomUUID());

        assertEquals(SettlementStatus.OPENED, settlement.status());
    }

    private static void assertEquals(Object expected, Object actual) {
        if (!expected.equals(actual)) {
            throw new AssertionError("Expected " + expected + " but got " + actual);
        }
    }

    private static void assertThrows(Class<? extends Throwable> expectedType, Runnable action) {
        try {
            action.run();
        } catch (Throwable actual) {
            if (expectedType.isInstance(actual)) {
                return;
            }
            throw new AssertionError("Expected " + expectedType.getSimpleName() + " but got " + actual.getClass().getSimpleName(), actual);
        }
        throw new AssertionError("Expected " + expectedType.getSimpleName() + " to be thrown");
    }
}

File: src/test/kotlin/com/onezero/oop/settlement/SettlementTest.kt
package com.onezero.oop.settlement

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class SettlementTest {
    @Test
    fun `정상 정산 전이를 허용한다`() {
        val settlement = Settlement()

        settlement.transitionTo(SettlementStatus.AGGREGATING)
        settlement.transitionTo(SettlementStatus.VALIDATING)
        settlement.transitionTo(SettlementStatus.CONFIRMED)

        assertEquals(SettlementStatus.CONFIRMED, settlement.status)
    }

    @Test
    fun `잘못된 전이를 차단한다`() {
        val settlement = Settlement()

        assertFailsWith<InvalidTransitionException> {
            settlement.transitionTo(SettlementStatus.CONFIRMED)
        }
    }

    @Test
    fun `확정 후 정정 요청을 허용한다`() {
        val settlement = Settlement()

        settlement.transitionTo(SettlementStatus.AGGREGATING)
        settlement.transitionTo(SettlementStatus.VALIDATING)
        settlement.transitionTo(SettlementStatus.CONFIRMED)
        settlement.transitionTo(SettlementStatus.CORRECTION_REQUESTED)

        assertEquals(SettlementStatus.CORRECTION_REQUESTED, settlement.status)
    }

    @Test
    fun `실패 상태에서 확정으로 돌아갈 수 없다`() {
        val settlement = Settlement()

        settlement.transitionTo(SettlementStatus.FAILED)

        assertFailsWith<InvalidTransitionException> {
            settlement.transitionTo(SettlementStatus.CONFIRMED)
        }
    }

    @Test
    fun `현재 상태를 조회할 수 있다`() {
        val settlement = Settlement()

        assertEquals(SettlementStatus.OPENED, settlement.status)
    }
}
